Update ratelimits configuration interface

alfrunes opened the pull request:
The behavior of the default rate limiting group can be confusing, so I thought it might be a good idea to remove it and replace it with an (almost) equivalent default config:
```yaml
# Replaces the `default` configuration block:
groups:
- name: default
  # ... (default paramters)
match:
- api_pattern: / # this pattern is a catch-all with lowest priority
  group_expression: "default"
```
The default is now to ignore ALL unmatched requests with a new option to reject them to make configuration more predictable.

alfrunes committed:
refactor(pkg): change the ratelimits configuration interface

Removed the default group configuration option since it is complicating
the interface. It is much cleaner to setup the default rule as a single
match and group rule:
```yaml
groups:
  - name: default
    # ...
match:
  - api_pattern: "/"
    event_expression: "default"
```

Changelog: None
Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
chore(pkg): Added ratelimit config parameter RejectUnmatched

Adds an option to reject requests which does not resolve to a ratelimit
configuration.

Changelog: None
Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
feat(useradm): Add rate limiting configuration for authenticated requests

Added the following configuration parameters:

```yaml
ratelimits:
  # auth configures ratelimits for authenticated requests.
  auth:
    # enable rate limiting also requires redis_connection_string to be effective.
    enable: false
    # reject_unmatched rejects requests that does not resolve to a
    # ratelimit group. That is, if either there's no api_pattern matching
    # the request or if the group_expression does not match a group.
    # Defaults to false - disable ratelimiting for unmatched requests.
    reject_unmatched: false
    # groups specify rate limiting groups that overrides the parameters in the
    # default group.
    groups:
        # name defines the name of the group. The name is used in
        # match.group_expression to match an api_pattern with a group.
      - name: default
        # interval is the time interval when the rate limiter resets.
        interval: 1m
        # quota is the number of requests allowed in an interval.
        quota: 300
        # event_expression is a go template for grouping requests.
        # The following attributes are available in the context:
        # Identity - contains a subset of the JWT claims:
        # .Subject  (jwt:"sub")          string
        # .Tenant   (jwt:"mender.tenant") string
        # .Plan     (jwt:"mender.plan")   string
        # .Addons   (jwt:"mender.addons") []struct{Enabled bool; Name string}
        # .IsUser   (jwt:"mender.user")   bool
        # .IsDevice (jwt:"mender.device") bool
        # .Trial    (jwt:"mender.trial")  bool
        event_expression: "{{with .Identity}}{{.Subject}}{{end}}"
        # More example groups:
    # - name: "example_slow_group_by_tenant"
    #   quota: 1
    #   interval: 30s
    #   event_expression: "{{with .Identity}}{{.Tenant}}{{end}}"
    # match specifies matching expressions for mapping API requests to rate
    # limiting groups.
    match:
        # api_pattern specifies an API path pattern as defined by http.ServeMux
        # https://pkg.go.dev/net/http#hdr-Patterns-ServeMux
      - api_pattern: /
        # group_expression defines  the group for this matching expression.
        # A group can be selected dynamically using Go templates or statically
        # with a literal string.
        # See group.event_expression for template context attributes.
        group_expression: "default"
        # More example match rules:
    # - api_pattern: /api/management/v1/useradm/expample/slow/api/
    #   group_expression: "example_slow_group_by_tenant"
    # - api_pattern: /api/management/v1/useradm/expample/slow/api/
    #   group_expression: >
    #     {{-if .Identity.Trial-}}
    #     example_slow_group_by_tenant
    #     {{-else-}}
    #     default
    #     {{-end-}}
```

Changelog: Commit
Ticket: MEN-7745
Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
feat(deviceauth): Add rate limiting configuration for authenticated requests

Added the following configuration parameters:

```yaml
ratelimits:
  # auth configures ratelimits for authenticated requests.
  auth:
    # enable rate limiting also requires redis_connection_string to be effective.
    enable: false
    # reject_unmatched rejects requests that does not resolve to a
    # ratelimit group. That is, if either there's no api_pattern matching
    # the request or if the group_expression does not match a group.
    # Defaults to false - disable ratelimiting for unmatched requests.
    reject_unmatched: false
    # groups specify rate limiting groups that overrides the parameters in the
    # default group.
    groups:
        # name defines the name of the group. The name is used in
        # match.group_expression to match an api_pattern with a group.
      - name: default
        # interval is the time interval when the rate limiter resets.
        interval: 1m
        # quota is the number of requests allowed in an interval.
        quota: 300
        # event_expression is a go template for grouping requests.
        # The following attributes are available in the context:
        # Identity - contains a subset of the JWT claims:
        # .Subject  (jwt:"sub")          string
        # .Tenant   (jwt:"mender.tenant") string
        # .Plan     (jwt:"mender.plan")   string
        # .Addons   (jwt:"mender.addons") []struct{Enabled bool; Name string}
        # .IsUser   (jwt:"mender.user")   bool
        # .IsDevice (jwt:"mender.device") bool
        # .Trial    (jwt:"mender.trial")  bool
        event_expression: "{{with .Identity}}{{.Subject}}{{end}}"
        # More example groups:
    # - name: "example_slow_group_by_tenant"
    #   quota: 1
    #   interval: 30s
    #   event_expression: "{{with .Identity}}{{.Tenant}}{{end}}"
    # match specifies matching expressions for mapping API requests to rate
    # limiting groups.
    match:
        # api_pattern specifies an API path pattern as defined by http.ServeMux
        # https://pkg.go.dev/net/http#hdr-Patterns-ServeMux
      - api_pattern: /
        # group_expression defines  the group for this matching expression.
        # A group can be selected dynamically using Go templates or statically
        # with a literal string.
        # See group.event_expression for template context attributes.
        group_expression: "default"
        # More example match rules:
    # - api_pattern: /api/devices/v1/deployments/expample/slow/api/
    #   group_expression: "example_slow_group_by_tenant"
    # - api_pattern: /api/devices/v2/devauth/expample/slow/api/
    #   group_expression: >
    #     {{-if .Identity.Trial-}}
    #     example_slow_group_by_tenant
    #     {{-else-}}
    #     default
    #     {{-end-}}
```

Changelog: Commit
Ticket: MEN-7744
Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

kjaskiewiczz reviewed (commented):
Is this one ready for review? (it doesn't pass static checks)
Is this part of https://northerntech.atlassian.net/browse/MEN-7744? (sorry, I was off recently and lost track of what's going on with rate limits)

alfrunes commented:
> Is this one ready for review? (it doesn't pass static checks) Is this part of https://northerntech.atlassian.net/browse/MEN-7744? (sorry, I was off recently and lost track of what's going on with rate limits)

I fixed the static check errors.
About the rate limiting stuff, I have to admit this turned out to be a bit of a mess. In this PR I'm trying to simplify the configuration interface a little by removing the confusing "default" limits and instead define the limits using the api patterns instead. I also added the "reject_unmatched" option since the replacement of the default is to accept requests that don't match a ratelimiting group.

kjaskiewiczz reviewed (approved):
looks good
